### PR TITLE
feat: Adding nrfcaching library

### PR DESCRIPTION
**Pull request by gatici**
Move the NRF caching library from https://github.com/omec-project/nrf/tree/master/nrfcache to this repository. Because all the NFs requires it and keeping the code in a common repository helps to mange the dependencies between NFs.

### CODE REVIEW

**Review by dariofaccin — APPROVED** [review]

File: nrfcache/mach_filters.go
Comment (dariofaccin) [inline nrfcache/mach_filters.go:1]: This file should be named `match_filters.go`. Otherwise looks good.

**Review by dariofaccin — CHANGES_REQUESTED** [review]

File: nrfcache/mach_filters.go
Comment (gatici) [reply nrfcache/mach_filters.go:1]: Renamed. Thanks for noticing.

**Review by dariofaccin — APPROVED** [review]

File: nrfcache/match_filters.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+					}
+				}
+			}
+		}
+
+		if dnnMatched == false {
+			return false, nil
+		}
+	}
+	fmt.Printf("SMF match found, nfInstance Id %v", profile.NfInstanceId)
```
Comment (gab-arrobo) [inline nrfcache/match_filters.go:115]:
```suggestion
	fmt.Printf("smf match found, nfInstance Id %v", profile.NfInstanceId)
```

File: nrfcache/match_filters.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+func MatchAusfProfile(profile *models.NfProfile, opts *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (bool, error) {
+	matchFound := true
+	if opts.Supi.IsSet() {
+		if profile.AusfInfo != nil && len(profile.AusfInfo.SupiRanges) > 0 {
+			matchFound = MatchSupiRange(opts.Supi.Value(), profile.AusfInfo.SupiRanges)
+		}
+	}
+	fmt.Printf("Ausf match found = %v", matchFound)
```
Comment (gab-arrobo) [inline nrfcache/match_filters.go:145]:
```suggestion
	fmt.Printf("ausf match found = %v", matchFound)
```

File: nrfcache/match_filters.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	return matchFound, nil
+}
+
+func MatchNssfProfile(profile *models.NfProfile, opts *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (bool, error) {
+	fmt.Println("Nssf match found ")
```
Comment (gab-arrobo) [inline nrfcache/match_filters.go:150]:
```suggestion
	fmt.Println("nssf match found ")
```

File: nrfcache/match_filters.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+		if profile.PlmnList != nil {
+			plmnMatchCount := 0
+
+			targetPlmnList := opts.TargetPlmnList.Value().([]string)
+			for _, targetPlmn := range targetPlmnList {
+				var plmn models.PlmnId
+				err := json.Unmarshal([]byte(targetPlmn), &plmn)
+
+				if err != nil {
+					fmt.Printf("Error Unmarshaling plmn : %+v", err)
```
Comment (gab-arrobo) [inline nrfcache/match_filters.go:166]:
```suggestion
					fmt.Printf("error unmarshaling plmn : %+v", err)
```

File: nrfcache/match_filters.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+			if profile.AmfInfo.GuamiList != nil {
+				guamiMatchCount := 0
+
+				guamiList := opts.Guami.Value().([]string)
+				for _, guami := range guamiList {
+					var guamiOpt models.Guami
+					err := json.Unmarshal([]byte(guami), &guamiOpt)
+
+					if err != nil {
+						fmt.Printf("Error Unmarshaling guami : %+v", err)
```
Comment (gab-arrobo) [inline nrfcache/match_filters.go:194]:
```suggestion
						fmt.Printf("error unmarshaling guami : %+v", err)
```

File: nrfcache/match_filters.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+		if opts.TargetNfInstanceId.IsSet() {
+			if profile.NfInstanceId != "" {
+				if profile.NfInstanceId != opts.TargetNfInstanceId.Value() {
+					return false, nil
+				}
+			}
+		}
+	}
+	fmt.Printf("Amf match found = %v", profile.NfInstanceId)
```
Comment (gab-arrobo) [inline nrfcache/match_filters.go:235]:
```suggestion
	fmt.Printf("amf match found = %v", profile.NfInstanceId)
```

File: nrfcache/match_filters.go [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+func MatchPcfProfile(profile *models.NfProfile, opts *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (bool, error) {
+	matchFound := true
+	if opts.Supi.IsSet() {
+		if profile.PcfInfo != nil && len(profile.PcfInfo.SupiRanges) > 0 {
+			matchFound = MatchSupiRange(opts.Supi.Value(), profile.PcfInfo.SupiRanges)
+		}
+	}
+	fmt.Printf("PCF match found = %v", matchFound)
```
Comment (gab-arrobo) [inline nrfcache/match_filters.go:246]:
```suggestion
	fmt.Printf("pcf match found = %v", matchFound)
```

File: nrfcache/match_filters.go [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+func MatchUdmProfile(profile *models.NfProfile, opts *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (bool, error) {
+	matchFound := true
+	if opts.Supi.IsSet() {
+		if profile.UdmInfo != nil && len(profile.UdmInfo.SupiRanges) > 0 {
+			matchFound = MatchSupiRange(opts.Supi.Value(), profile.UdmInfo.SupiRanges)
+		}
+	}
+	fmt.Printf("UDM match found = %v", matchFound)
```
Comment (gab-arrobo) [inline nrfcache/match_filters.go:257]:
```suggestion
	fmt.Printf("udm match found = %v", matchFound)
```

File: nrfcache/nrfcache_test.go [hunk 9]
```diff
@@ -0,0 +1,902 @@
...
+}
+
+func MarshToJsonString(v interface{}) (result []string) {
+	types := reflect.TypeOf(v)
+	val := reflect.ValueOf(v)
+	if types.Kind() == reflect.Slice {
+		for i := 0; i < val.Len(); i++ {
+			tmp, err := json.Marshal(val.Index(i).Interface())
+			if err != nil {
+				fmt.Printf("Marshal error: %+v", err)
```
Comment (gab-arrobo) [inline nrfcache/nrfcache_test.go:446]:
```suggestion
				fmt.Printf("marshal error: %+v", err)
```

File: nrfcache/nrfcache_test.go [hunk 10]
```diff
@@ -0,0 +1,902 @@
...
+			}
+
+			result = append(result, string(tmp))
+		}
+	} else {
+		tmp, err := json.Marshal(v)
+		if err != nil {
+			fmt.Printf("Marshal error: %+v", err)
```
Comment (gab-arrobo) [inline nrfcache/nrfcache_test.go:454]:
```suggestion
			fmt.Printf("marshal error: %+v", err)
```

File: nrfcache/nrfcache.go [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+	mutex               sync.Mutex
+}
+
+func (c *NrfMasterCache) GetNrfCacheInstance(targetNfType models.NfType) *NrfCache {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	cache, exists := c.nfTypeToCacheMap[targetNfType]
+	if exists == false {
+		fmt.Printf("Creating cache for nftype %v", targetNfType)
```
Comment (gab-arrobo) [inline nrfcache/nrfcache.go:300]:
```suggestion
		fmt.Printf("creating cache for nftype %v", targetNfType)
```

File: nrfcache/nrfcache.go [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+func (c *NrfCache) handleLookup(nrfUri string, targetNfType, requestNfType models.NfType, param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (models.SearchResult, error) {
+	var searchResult models.SearchResult
+	var err error
+
+	c.mutex.RLock()
+	searchResult.NfInstances = c.get(param)
+	c.mutex.RUnlock()
+
+	if len(searchResult.NfInstances) == 0 {
+		fmt.Printf("Cache miss for nftype %s", targetNfType)
```
Comment (gab-arrobo) [inline nrfcache/nrfcache.go:150]:
```suggestion
		fmt.Printf("cache miss for nftype %s", targetNfType)
```

File: nrfcache/nrfcache.go [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+
+	var searchResult models.SearchResult
+	var err error
+
+	c := masterCache.GetNrfCacheInstance(targetNfType)
+	if c != nil {
+		searchResult, err = c.handleLookup(nrfUri, targetNfType, requestNfType, param)
+	} else {
+		fmt.Println("SearchNFInstances nrf cache")
+
```
Comment (gab-arrobo) [inline nrfcache/nrfcache.go:358]: Why did the message changed. In the `nrf`, the message says "Failed to find cache for nftype". Should not the message here be kind of similar? or am I overlooking somthing?
```suggestion
		fmt.Println("failed to find cache for nftype")
```

File: nrfcache/match_filters.go (hunk 1 above)
Comment (gatici) [reply nrfcache/match_filters.go:115]: Done

File: nrfcache/match_filters.go (hunk 2 above)
Comment (gatici) [reply nrfcache/match_filters.go:145]: Done

Comment (gatici) [reply nrfcache/match_filters.go]: Done

File: nrfcache/match_filters.go (hunk 3 above)
Comment (gatici) [reply nrfcache/match_filters.go]: Done

File: nrfcache/match_filters.go (hunk 4 above)
Comment (gatici) [reply nrfcache/match_filters.go]: Done

File: nrfcache/match_filters.go (hunk 5 above)
Comment (gatici) [reply nrfcache/match_filters.go]: Done

File: nrfcache/match_filters.go (hunk 6 above)
Comment (gatici) [reply nrfcache/match_filters.go]: Done

File: nrfcache/match_filters.go (hunk 7 above)
Comment (gatici) [reply nrfcache/match_filters.go]: Done

File: nrfcache/nrfcache_test.go (hunk 9 above)
Comment (gatici) [reply nrfcache/nrfcache_test.go]: Done

File: nrfcache/nrfcache_test.go (hunk 10 above)
Comment (gatici) [reply nrfcache/nrfcache_test.go]: Done

File: nrfcache/nrfcache.go (hunk 11 above)
Comment (gatici) [reply nrfcache/nrfcache.go]: Done

File: nrfcache/nrfcache.go (hunk 12 above)
Comment (gatici) [reply nrfcache/nrfcache.go]: Done

File: nrfcache/nrfcache.go (hunk 13 above)
Comment (gatici) [reply nrfcache/nrfcache.go]: This is my mistake. The original log message is used.

File: nrfcache/match_filters.go (hunk 8 above)
Comment (gatici) [reply nrfcache/match_filters.go]: Done

**Review by gab-arrobo — APPROVED** [review]
+1

**Pull request merged**